### PR TITLE
[REVIEW] Improve reading for: SELECT * FROM table LIMIT N

### LINKED ISSUE #623: SELECT * with LIMIT loads up all the partitions
When running a `select * from table limit n` it scans all the files as opposed to just returning the first n rows it finds.

This isn't an issue on datasets in memory, but on large datasets this can make it very difficult just to see what the data looks like.

### PR DESCRIPTION
**Pull request by Christian8491**
Closes #623 
When the SQL statement is as simple as:
`SELECT * FROM table LIMIT N`

It's not necessary to read all the files to show only N values.

### CODE REVIEW

**Comment by williamBlazing** [comment]
This implementation is a very very targetted optimization. 
A more generic optimization, which would be better and handle all sorts of other situations that would be useful it to add a mechanism to the engine that would allow you to terminate execution of a query. This way, then limit is applied and it detects that all the necessary rows are added to the output are done, it just terminates the query. This same mechanism can then be leveraged for much more complex queries. Or to allow a user, somehow to terminate a query early. Or if we implement a heartbeat system where we notice one node going down, we can terminate the query early as well.

In general, i am ok with this optmization, but we might want to get rid of it once we implement a mechanism for terminating a query

File: engine/src/execution_graph/logic_controllers/taskflow/graph.cpp [hunk 1]
```diff
@@ -250,6 +251,28 @@ namespace cache {
...
+	void graph::check_how_data_looks() {
+		auto first_iterator = container_.begin(); // points to null
+		first_iterator++;
+		int32_t min_index_valid = first_iterator->first;
+		size_t total_kernels = container_.size(); 
+
+		if (total_kernels == 4) { // null, TableScanKernel, LimitKernel, OutputKernel
+			if ( get_node(min_index_valid )->expression == "OutputKernel" &&
+				 get_node(min_index_valid + 1)->get_type_id() == kernel_type::LimitKernel &&
+			 	 get_node(min_index_valid + 2)->get_type_id() == kernel_type::TableScanKernel )
```
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/taskflow/graph.cpp:265]: this should be TableScan OR BindableTableScan

File: engine/src/execution_graph/logic_controllers/taskflow/graph.cpp [hunk 2]
```diff
@@ -250,6 +251,28 @@ namespace cache {
 			return std::set<graph::Edge>();
 		}
 	}
+
+	// This function will work when the Relational Algebra only contains: LogicalTableScan and LogicalLimit
+	void graph::check_how_data_looks() {
```
[the remark this answers is not in the document]
Comment (Christian8491) [reply engine/src/execution_graph/logic_controllers/taskflow/graph.cpp]: done!

File: engine/src/execution_graph/logic_controllers/taskflow/graph.cpp (hunk 1 above)
Comment (Christian8491) [reply engine/src/execution_graph/logic_controllers/taskflow/graph.cpp]: Done!

**Comment by kharoc** [comment]
1 RAL

========================================================
FAILED TESTS                                                                                                                                                                                                
========================================================                                                                                                                                                    
                        TestGroup InputType Result                                           TestId
0  Bindable Alias                  orc       Fail   TEST_09                                        
1  Date                            orc       Fail   TEST_01                                        
2  Simple Distribution From Local  orc       Fail   TEST_22,TEST_43,TEST_46,TEST_49                
3  TPCH Queries                    orc       Fail   TEST_03,TEST_05,TEST-08,TEST_10,TEST_14,TEST_18
4  Timestamp                       orc       Fail   TEST_02                                        
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1269/1282
FAILED: 13/1282
CRASH: 0/1282
TOTAL: 1282
MAX DELTA: 48.0
*******************************************************************************
Aggregations without group by:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Coalesce:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Column Basis:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Common Table Expressions:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Count without group by:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Date:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Timestamp:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Full outer join:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Group by:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Group by without aggregations:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Inner join:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Left outer join:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Non-EquiJoin Queries:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Nested Queries:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Order by:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Predicates With Nulls:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Simple String:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Unary ops:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Unify Tables:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Union:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Limit:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Where clause:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Bindable Alias:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Boolean:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Case:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Cast:   Start Mem: 8119.875   End Mem: 8119.875   Diff: 0.0
Concat:   Start Mem: 8119.875   End Mem: 8167.875   Diff: 48.0
Literal:   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0
Dir:   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0
File System Local:   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0
File System S3:   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0
File System Google Storage :   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0
Like:   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0
Substring:   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0
Wild Card:   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0
TPCH Queries:   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0
Round:   Start Mem: 8167.875   End Mem: 8167.875   Diff: 0.0

2 Worker's

========================================================
FAILED TESTS                                                                                                                                                                                                
========================================================                                                                                                                                                    
                        TestGroup InputType Result                                           TestId
0  Bindable Alias                  orc       Fail   TEST_09                                        
1  Date                            orc       Fail   TEST_01                                        
2  Simple Distribution From Local  orc       Fail   TEST_22,TEST_43,TEST_46,TEST_49                
3  TPCH Queries                    orc       Fail   TEST_03,TEST_05,TEST-08,TEST_10,TEST_14,TEST_18
4  Timestamp                       orc       Fail   TEST_02                                        
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1508/1521
FAILED: 13/1521
CRASH: 0/1521
TOTAL: 1521
MAX DELTA: 457.0
*******************************************************************************
Aggregations without group by:   Start Mem: 868.875   End Mem: 1325.875   Diff: 457.0
Coalesce:   Start Mem: 1325.875   End Mem: 1349.875   Diff: 24.0
Common Table Expressions:   Start Mem: 1349.875   End Mem: 1343.875   Diff: -6.0
Count without group by:   Start Mem: 1343.875   End Mem: 1343.875   Diff: 0.0
Date:   Start Mem: 1343.875   End Mem: 1361.875   Diff: 18.0
Timestamp:   Start Mem: 1361.875   End Mem: 1361.875   Diff: 0.0
Full outer join:   Start Mem: 1361.875   End Mem: 1361.875   Diff: 0.0
Group by:   Start Mem: 1361.875   End Mem: 1373.875   Diff: 12.0
Group by without aggregations:   Start Mem: 1373.875   End Mem: 1373.875   Diff: 0.0
Inner join:   Start Mem: 1373.875   End Mem: 1379.875   Diff: 6.0
Left outer join:   Start Mem: 1379.875   End Mem: 1379.875   Diff: 0.0
Non-EquiJoin Queries:   Start Mem: 1379.875   End Mem: 1389.875   Diff: 10.0
Nested Queries:   Start Mem: 1389.875   End Mem: 1389.875   Diff: 0.0
Order by:   Start Mem: 1389.875   End Mem: 1389.875   Diff: 0.0
Predicates With Nulls:   Start Mem: 1389.875   End Mem: 1395.875   Diff: 6.0
Simple String:   Start Mem: 1395.875   End Mem: 1395.875   Diff: 0.0
Unary ops:   Start Mem: 1669.875   End Mem: 1667.875   Diff: -2.0
Unify Tables:   Start Mem: 1667.875   End Mem: 1667.875   Diff: 0.0
Union:   Start Mem: 1667.875   End Mem: 1627.875   Diff: -40.0
Limit:   Start Mem: 1627.875   End Mem: 1415.875   Diff: -212.0
Where clause:   Start Mem: 1415.875   End Mem: 1415.875   Diff: 0.0
Bindable Alias:   Start Mem: 1415.875   End Mem: 1427.875   Diff: 12.0
Boolean:   Start Mem: 1427.875   End Mem: 1427.875   Diff: 0.0
Case:   Start Mem: 1427.875   End Mem: 1435.875   Diff: 8.0
Cast:   Start Mem: 1435.875   End Mem: 1445.875   Diff: 10.0
Concat:   Start Mem: 1445.875   End Mem: 1567.875   Diff: 122.0
Literal:   Start Mem: 1567.875   End Mem: 1567.875   Diff: 0.0
Dir:   Start Mem: 1567.875   End Mem: 1871.875   Diff: 304.0
File System Local:   Start Mem: 1871.875   End Mem: 1611.875   Diff: -260.0
File System S3:   Start Mem: 1611.875   End Mem: 1611.875   Diff: 0.0
File System Google Storage :   Start Mem: 1611.875   End Mem: 1611.875   Diff: 0.0
Like:   Start Mem: 1611.875   End Mem: 1837.875   Diff: 226.0
Substring:   Start Mem: 1777.875   End Mem: 1755.875   Diff: -22.0
Wild Card:   Start Mem: 1755.875   End Mem: 1769.875   Diff: 14.0
TPCH Queries:   Start Mem: 1769.875   End Mem: 1795.875   Diff: 26.0
Round:   Start Mem: 1795.875   End Mem: 1795.875   Diff: 0.0

File: engine/src/execution_graph/logic_controllers/taskflow/kernel.h [hunk 3]
```diff
@@ -125,6 +127,10 @@ class kernel {
 	std::shared_ptr<graph> query_graph;
 	std::shared_ptr<Context> context;
 
+	// useful when the Algebra Relacional only contains: LogicalTableScan and LogicalLimit
+	std::atomic<bool> has_limit_;
```
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/taskflow/kernel.h:131]: why are these atomic? these are only set once in a single threaded part of the code and then they are never modified, right?

Comment (Christian8491) [reply engine/src/execution_graph/logic_controllers/taskflow/kernel.h]: Yep, you are right. It's only set once and never modified. I just remove the `std::atomic` part . 
Also I tested this feature on the dgx1.

**Comment by kharoc** [comment]
1 RAL

========================================================
FAILED TESTS                                                                                                                                 
========================================================                                                                                     
Empty DataFrame
Columns: [index, TestId]
Index: []
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1290/1290
FAILED: 0/1290
CRASH: 0/1290
TOTAL: 1290
MAX DELTA: 48.0
*******************************************************************************
Aggregations without group by:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Coalesce:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Column Basis:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Common Table Expressions:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Count without group by:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Date:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Timestamp:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Full outer join:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Group by:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Group by without aggregations:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Inner join:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Left outer join:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Non-EquiJoin Queries:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Nested Queries:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Order by:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Predicates With Nulls:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Simple String:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Unary ops:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Unify Tables:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Union:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Limit:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Where clause:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Bindable Alias:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Boolean:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Case:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Cast:   Start Mem: 8059.875   End Mem: 8059.875   Diff: 0.0
Concat:   Start Mem: 8059.875   End Mem: 8107.875   Diff: 48.0
Literal:   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0
Dir:   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0
File System Local:   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0
File System S3:   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0
File System Google Storage :   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0
Like:   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0
Substring:   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0
Wild Card:   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0
TPCH Queries:   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0
Round:   Start Mem: 8107.875   End Mem: 8107.875   Diff: 0.0

2 Worker's

========================================================
FAILED TESTS                                                                                                                       
========================================================                                                                           
Empty DataFrame
Columns: [index, TestId]
Index: []
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1531/1531
FAILED: 0/1531
CRASH: 0/1531
TOTAL: 1531
MAX DELTA: 397.0
*******************************************************************************
Aggregations without group by:   Start Mem: 748.875   End Mem: 1145.875   Diff: 397.0
Coalesce:   Start Mem: 1145.875   End Mem: 1169.875   Diff: 24.0
Common Table Expressions:   Start Mem: 1169.875   End Mem: 1161.875   Diff: -8.0
Count without group by:   Start Mem: 1161.875   End Mem: 1161.875   Diff: 0.0
Date:   Start Mem: 1161.875   End Mem: 1171.875   Diff: 10.0
Timestamp:   Start Mem: 1171.875   End Mem: 1171.875   Diff: 0.0
Full outer join:   Start Mem: 1171.875   End Mem: 1171.875   Diff: 0.0
Group by:   Start Mem: 1171.875   End Mem: 1181.875   Diff: 10.0
Group by without aggregations:   Start Mem: 1181.875   End Mem: 1181.875   Diff: 0.0
Inner join:   Start Mem: 1181.875   End Mem: 1189.875   Diff: 8.0
Left outer join:   Start Mem: 1189.875   End Mem: 1189.875   Diff: 0.0
Non-EquiJoin Queries:   Start Mem: 1189.875   End Mem: 1209.875   Diff: 20.0
Nested Queries:   Start Mem: 1209.875   End Mem: 1209.875   Diff: 0.0
Order by:   Start Mem: 1209.875   End Mem: 1209.875   Diff: 0.0
Predicates With Nulls:   Start Mem: 1209.875   End Mem: 1215.875   Diff: 6.0
Simple String:   Start Mem: 1215.875   End Mem: 1215.875   Diff: 0.0
Unary ops:   Start Mem: 1489.875   End Mem: 1487.875   Diff: -2.0
Unify Tables:   Start Mem: 1487.875   End Mem: 1487.875   Diff: 0.0
Union:   Start Mem: 1487.875   End Mem: 1447.875   Diff: -40.0
Limit:   Start Mem: 1447.875   End Mem: 1235.875   Diff: -212.0
Where clause:   Start Mem: 1235.875   End Mem: 1235.875   Diff: 0.0
Bindable Alias:   Start Mem: 1235.875   End Mem: 1245.875   Diff: 10.0
Boolean:   Start Mem: 1245.875   End Mem: 1245.875   Diff: 0.0
Case:   Start Mem: 1245.875   End Mem: 1253.875   Diff: 8.0
Cast:   Start Mem: 1253.875   End Mem: 1261.875   Diff: 8.0
Concat:   Start Mem: 1261.875   End Mem: 1379.875   Diff: 118.0
Literal:   Start Mem: 1379.875   End Mem: 1379.875   Diff: 0.0
Dir:   Start Mem: 1379.875   End Mem: 1681.875   Diff: 302.0
File System Local:   Start Mem: 1681.875   End Mem: 1433.875   Diff: -248.0
File System S3:   Start Mem: 1433.875   End Mem: 1433.875   Diff: 0.0
File System Google Storage :   Start Mem: 1433.875   End Mem: 1433.875   Diff: 0.0
Like:   Start Mem: 1433.875   End Mem: 1661.875   Diff: 228.0
Substring:   Start Mem: 1473.875   End Mem: 1449.875   Diff: -24.0
Wild Card:   Start Mem: 1449.875   End Mem: 1463.875   Diff: 14.0
TPCH Queries:   Start Mem: 1463.875   End Mem: 1473.875   Diff: 10.0
Round:   Start Mem: 1473.875   End Mem: 1473.875   Diff: 0.0

**Pull request merged**